azureDatabaseUtils/src/main/java/bio/terra/workspace/azureDatabaseUtils/database/DatabaseService.java

```diff
@@ -0,0 +1,72 @@
+package bio.terra.workspace.azureDatabaseUtils.database;
+
+import bio.terra.workspace.azureDatabaseUtils.validation.Validator;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class DatabaseService {
+  private static final Logger logger = LoggerFactory.getLogger(DatabaseService.class);
+  private final DatabaseDao databaseDao;
+  private final Validator validator;
+
+  @Autowired
+  public DatabaseService(DatabaseDao databaseDao, Validator validator) {
+    this.databaseDao = databaseDao;
+    this.validator = validator;
+  }
+
+  // TODO: remove with https://broadworkbench.atlassian.net/browse/WOR-1165
+  public void createDatabaseWithManagedIdentity(
+      String newDbName, String newDbUserName, String newDbUserOid) {
+    validator.validateDatabaseNameFormat(newDbName);
+    validator.validateRoleNameFormat(newDbUserName);
+    validator.validateUserOidFormat(newDbUserOid);
+
+    logger.info(
+        "Creating database {} with user {} and OID {}", newDbName, newDbUserName, newDbUserOid);
+
+    databaseDao.createDatabase(newDbName);
+    databaseDao.createRoleForManagedIdentity(newDbUserName, newDbUserOid);
+    databaseDao.grantAllPrivileges(newDbUserName, newDbName);
+    databaseDao.revokeAllPublicPrivileges(newDbName);
+  }
+
+  public void createDatabaseWithDbRole(String newDbName) {
+    validator.validateDatabaseNameFormat(newDbName);
+
+    logger.info("Creating database {} with db role of same name", newDbName);
+
+    databaseDao.createDatabase(newDbName);
+    databaseDao.createRole(newDbName);
+    databaseDao.grantAllPrivileges(newDbName, newDbName);
+    databaseDao.revokeAllPublicPrivileges(newDbName);
+  }
+
+  public void createLoginRole(
+      String newDbUserName, String newDbUserOid, Set<String> databaseNames) {
+    validator.validateRoleNameFormat(newDbUserName);
+    validator.validateUserOidFormat(newDbUserOid);
+    databaseNames.forEach(validator::validateDatabaseNameFormat);
+
+    logger.info(
+        "Creating login role {} with OID {} for databases {}",
+        newDbUserName,
+        newDbUserOid,
+        databaseNames);
+
+    databaseDao.createRoleForManagedIdentity(newDbUserName, newDbUserOid);
```

Comment:
I'm not understanding how this knows to grant the login role-- it seems to just supply the username (which I'm thinking of as "fred") and an Oid (what is this?). I would expect to see something that specifies "login" as the role.

Perhaps things will become clearer as I read more. But if the userName is supposed to be the role, which is it not called a role? And if this method is just meant to grant the login role, why does it allow any username/role to be passed in?

Reply:
Having read further, this is used by "CreateUserRunner", and deleteLoginRole is used by "DeleteUserRunner". So why aren't these methods just `createUser` and `deleteUser`?

I'm still unclear about usernames vs. roles, and how we know which roles are supported.

Reply:
So this is a namespace role, since "Namespace roles are specific to a namespace, can login to the database and have access to 1 or more database roles.". But I find the code really difficult to understand with regard to namespace roles, database roles, and usernames. I think I will need for us to talk through this tomorrow because even having read the spec and gone through all the code, I'm still having trouble understanding how it all fits together.

Reply:
I also would like to talk through how the private database path works.